### DISCUSSION TITLE
Add logging to file on mass storage (at least USB), screenshots, etc.

### DISCUSSION THREAD

**Discussion by OneOfMany07** [memtest86plus/memtest86plus#164] [Enhancement]
Would it be useful to add code to create screenshots at a button press for user generated debug information?  Like if they don't have a phone handy.  Or want to automate the process of recording progress (the app could trigger output periodically in case it reboots).

Or even just dump the currently displayed test to a file for later use.  I assume that'd be simpler to do without adding lots of non-memory testing code, though I've never tried to write screenshots through code before.  Also uses less space if they're on a tiny old USB stick.

Both would be better if they made sure not to overwrite existing files, like by including the current time + date in the filename.  Maybe some other identifiable info too, like CPU + RAM stats or MAC address for uniqueness.

**Comment by debrouxl** [comment]
Thanks for the brainstorming work.

The ability to save data - screenshots, error reports, debugging information, etc. - would depend on support for USB MSD + one filesystem (especially FAT16/FAT32), which is a wish list item representing a sizable amount of work, mentioned e.g. in #138 . It doesn't seem to have a dedicated tracking discussion / issue for now.

MAC address information is nice for uniqueness, but it would require adding drivers for a wide range of NICs, some of which might require loading firmware beforehand, and represent ongoing maintenance to add support for new NICs... that's even more work, hard to justify for a non-commercial project without full-time paid developers :)

**Comment by x86fr** [comment]
> Would it be useful to add code to create screenshots at a button press for user generated debug information? Like if they don't have a phone handy. Or want to automate the process of recording progress (the app could trigger output periodically in case it reboots).

There is a misunderstanding about how Memtest86+ works. Memtest86+ does not rely on an underlying framework or operating system. It is an operating system on its own, not just an UEFI application like Passmark's commercial clone or any Windows or Linux software. This allows Memtest86+ to test 100% of the memory, without leaving some memory untested like the one needed by the framework or the OS.

To write any data to an USB Flash Drive, we can't just ask an underlying library. There is no library. We must code the whole low-level driver to access the USB Drive, then code the filesystem implementation to read or write data. This is a LOT of code and work.

**Reply by martinwhitaker** [reply]
Also we would almost certainly want to support hot-plugging the USB drive. This would mean more memory was reserved for use by the USB controllers and hence not covered by the memory tests.

**Comment by ghchris2021** [comment]
I agree with the sentiments of the ideas to provide (somehow / various possible ways of) logging, I was about to suggest similarly myself but searched for similar enhancement requests before doing so so I'll comment here.

* I think logging outputs might be possible / interesting to support by optionally emitting them from a network interface -- certainly common / simple types of wired ethernet NICs are usually pretty easy to send packets to even without any significant driver / networking code (e.g. just using the register interface and say open source minimal driver code maybe from U-Boot, RTOS, BSD, DOS, whatever).

* Obviously UART ports are possible low hanging fruit for possible log output interfaces.  Beyond the basic dedicated UART choices there are also plenty of USB-to-UART interface ICs like some from FTDI, Infineon/Cypress, Microchip which have well defined host USB side device class / interfaces and aren't so hard to talk to from embedded / bare metal code and if it's somehow possible to get them initialized at the BIOS/UEFI level then maybe even simpler if the USB HUBS / enumeration etc. is all done before the memtest runs so it can just emit characters into a buffer and not much else.

* Most mass storage disk-like devices are actually possibly extremely simple to talk to from bare metal / embedded code and that can be done even without a complex driver and possibly without a file system layer at all.  ATA, SATA, NVME, all potentially worth considering.

* In many common cases memtest may be loaded running from a BIOS/UEFI supported USB-mass storage flash drive, so in one simple possibility one might have a fixed partition of whatever 1GBy dedicated to booting / storing memtest, then the entire rest of that storage device could optionally be used to write log data if so configured / desired and presumably the BIOS/UEFI has already initialized the needed hubs / device etc. so maybe simplified acess to it via API could be possible to write sectors and that'd be sufficient if not desiring / implementing a file system for the log portion.
  
I understand this is running sort-of bare-metal so it's not completely trivial to "just use a file system" without an OS etc.
But to address that obstacle (a bit) I'll mention this -- I have worked with a lot of low level / bare metal systems over the years and sometimes one can have a certain feature despite making compromises for expediency in a constrained environment.  If one CAN writre logging to a designated USB mass storage or ATAPI or NVME or SATA or whatever drive then one doesn't HAVE to use either an OS or file-system -- for instance look at the classic UNIX "tar" utility that just writes records (originally to sequential output magnetic tape) one sector after another on the raw disk, no file system and only when the data is retrieved (under some OS) are the records scanned / interpreted by scanning from the start until the last valid one in order.  Similarly "syslog" or similar logging frameworks just write out serial data records one line / record after another and that works even on a character device (UART) or on a NIC (log packet 1, log packet 2, ...) each containing one more or less log line/record.  

So one can "stream" log lines / records to block devices (one sector / block = one record or blocking can be used to combine or aggregate shorter / longer records if desired).

One can just simply send IP/UDP packets out each containing one or more log lines / records to any NIC interface, no significant OS / driver needed as long as one can send a packet.  One can send a purely unidirectional no handshake needed packet such as that without any significant protocol overhead besides a few fixed content header fields that can be spit out from a C-struct.
https://en.wikipedia.org/wiki/User_Datagram_Protocol#UDP_datagram_structure
https://en.wikipedia.org/wiki/Internet_Protocol_version_4#Header
https://en.wikipedia.org/wiki/Syslog
https://datatracker.ietf.org/doc/html/rfc5424
https://datatracker.ietf.org/doc/html/rfc5426

One can even use source address 0.0.0.0  and broadcast destination address if desired though since there is a basic configuration TUI it doesn't seem problematic to manually let IP/UDP source/destination addresses be specified if desired.
One could even send out raw ethernet (non IP) frames and someone interested could capture them easily enough with wireshark or similar if they want.
If one wants to invest slightly more work there are stanrards (IETF/RFCs I think) for RSYSLOG / remote system log output protocols for outputting log data that will be compatible to be received by a standard UNIX network log collector daemon running on some LAN host in the same way log messages are emitted/collected by servers / workstations when they're up and emitting logs to a collection host.

On systems that support such I would guess that maybe somewhere within the IPMI / DMI / SMBIOS sorts of facilities that commonly exist there would be some kind of standardized API on supporting systems to capture / store / transmit logging information from system diagnostics / system test codes such as memtest.  If present I suppose the interfaces have to be pretty simple programmatically.  And if possible to use them then I assume standard COTS DMI / IPMI / whatever tools could then be used to read / receive / display / process such log data.

And although it's certainly a diferent operating environment entirely I simply wonder how hard / interesting it might be to make something like memtest run (as a variant tool or whatever) inside / under some already existing low level micro-framework such as a microkernel, U-BOOT, a RTOS (freertos, zephyr, rtems, minix, whatever) so that the memory test C/ASM/whatever could be given exclusive access to all or almost all physical RAM and yet there could potentially be just enough platform based micro-driver support to facilitate optionally talking to USB-serial, ATA, SATA, NVME, USB-mass-storage, wired-ethernet, ... devices / interfaces.

Looking at my current computer CPU capabilities --

* Almost all of the first SEVERAL generations of OS I've used ran in whole-system RAM sizes equal to or less than the 512 kBy of L2 cache my CPU has.  So really it seems not impossible for a RTOS, microkernel, DOS, OS/2, CPM, RSX, very "large" / capable test utility / whatever to be able to live ~entirely even out of CPU cache and not need significant amounts of run-time RAM (permitting it all to be tested and error prone).

* Looking at the 64MBy size of the CPU L3 cache, that's more than the total system RAM I've seen used on even much more significantly capable embedded OS environments e.g. PicoBSD "It runs on a minimum 386SX with 8 MB of RAM", MINIX, LINUX, RTOSs, ...

**Comment by debrouxl** [comment]
Thanks for your input.
Several technical notes to explain why memtest86+ is a custom code base which does everything on its own:
* some BIOS implementations do not initialize hardware properly, so kernels, which aim at reliable operation, usually have to redo it;
* UEFI boot services are a non-starter for memtest86+: we strive at providing the same feature set on legacy BIOS and UEFI BIOS, and worse, UEFI boot services permanently reserve memory which can't be tested, sometimes significant amounts thereof (dozens or hundreds of MBs, per another issue / discussion in this repo), whereas memtest86+ strives at testing the entire memory range, besides merely several 4 KB pages used for internal data structures: barriers, USB controller handling.
* AFAIK, most existing kernels permanently reserve a memory range and don't move themselves over time, whereas memtest86+ constantly relocates itself in order to be able to test the entire memory range (minus several 4 KB pages, again). The kernels which do not relocate themselves can't be used directly, and may require nontrivial modifications to fit this use case.
